### DISCUSSION TITLE
LivewireManager::$isLivewireRequestTestingOverride causing subsequent tests to fail

### DISCUSSION THREAD

**Discussion by fuelingtheweb** [livewire/livewire#3120] [7. Bugs (archived)]
### Description
With Livewire v2.3.9 and above, running my entire test suite causes tests that rely on the ConvertEmptyStringsToNull middleware to fail when they run after a Livewire test that sets data.

```
        Livewire::actingAs($admin)
            ->test(FilterUsers::class)
            ->set('sort', 'first_name')
            ->assertViewHas('users');
```

All tests are green when `->set()` is omitted, but tests that rely on ConvertEmptyStringsToNull fail when `->set()` is included.

When I remove `LivewireManager::$isLivewireRequestTestingOverride = true;` from `TestableLivewire@pretendWereSendingAComponentUpdateRequest()`, all tests pass. I'm not sure if anything else is relying on that code. Can that safely be removed?

**Context**
- Livewire version: >=2.3.9
- Laravel version: 8.27.0

**Comment by austenc** [comment]
Can you please provide more context so we can reproduce this without guessing? The test that fails, the livewire component, etc...

An example repository would be super helpful here. Thanks!

**Comment by fuelingtheweb** [comment]
@austenc Here's a sample Laravel application that replicates the issue: https://github.com/benholmen/livewire-test-issue

If you run the entire test suite, `tests/Feature/OtherTest.php@testShouldPassButFailsWhenAlsoTestingLivewire` fails because `Livewire@set` in `tests/Feature/LivewireTest.php@testSettingPropertyBreaksOtherTests` is disabling middleware. When run in isolation, all tests pass.

If Livewire testing needs to disable middleware, could it just reset the middleware when the Livewire test is finished?

**Comment by aleahy** [comment]
I've been finding the same thing. As @fuelingtheweb said, it only happens if the livewire version installed is 2.3.9 or greater.

The livewire component is not complicated - it just has a single property.
I've also created a separate, unrelated Post model with a controller to store new posts.

Here is an example of a livewire test that causes the fail:
```
public function test_it_can_update_value()
{
    Livewire::test(Something::class)
        ->set('textValue', 'Some value')
        ->assertViewHas('Some value');
}
```

Here is an example of a test that fails:
```
public function test_a_post_can_be_created_with_no_url()
{
    $this->post(route('posts.store'), [
            'title' => 'Some title',
            'url' => '',
            'message' => 'Some message',
        ])->assertSuccessful();

        $this->assertDatabaseHas('posts', [
            'title' => 'Some title',
            'url' => null,
            'message' => 'Some message'
        ]);
}
```

The reported reason for the test failing is that url has an empty string in the database, which suggests the middleware `ConvertEmptyStringsToNull` is being ignored.

Run independently, both tests pass.

**Comment by austenc** [comment]
Thanks folks, I'll check this out tomorrow!

**Comment by benholmen** [comment]
Thanks Austen!

**Comment by tomwelch** [comment]
Hi all,

I've stumbled across this too. As a workaround at the moment I'm resetting `$isLivewireRequestTestingOverride` to `false` in the `tearDown` method on any test class that interacts with Livewire's test helpers.

```php
protected function tearDown(): void
{
    parent::tearDown();

    LivewireManager::$isLivewireRequestTestingOverride = false;
}
```

Resetting after every test may be a bit excessive if your test class only tests a Liverwire component, at which point you can reset in the `tearDownAfterClass` method.

```php
public static function tearDownAfterClass(): void
{
    LivewireManager::$isLivewireRequestTestingOverride = false;
}
```

**Comment by ThibaudDauce** [comment]
Had the same problem today. Same solution as @tomwelch (I added a `LivewireManager::$isLivewireRequestTestingOverride = false;` in my `setUp()`.

**Comment by nvahalik** [comment]
We just ran into this problem as well. @tomwelch's fix works as expected!

**Comment by GC-Mark** [comment]
We've just come across this issue, @tomwelch solution works for now, but hopefully this will be fixed internally at some point 👍

**Comment by lava83** [comment]
Yes, I know the topic is almost half a year old, but I would still be interested to know why here: https://github.com/livewire/livewire/blob/fc8c315babf8d42c2a25f207b3931b1bc8eb5e70/src/LivewireServiceProvider.php#L77..L83 `TrimStrings::class` and `ConvertEmptyStringsToNull::class` - are excluded?

**Reply by joshhanley** [reply]
@lava83 they're excluded because Livewire's request is different to a standard Laravel request. So those two middleware would change properties that shouldn't be changed in a Livewire request, as such need to be removed/ excluded from running.

Hope this helps!

**Reply by JayBizzle** [reply]
Thanks @joshhanley for the info, but any thoughts on how we can go about fixing the main issue in this thread without having to resort to the solution suggested by @tomwelch

**Reply by joshhanley** [reply]
@JayBizzle hmm at a quick glance, I'm not sure. I don't know how we could track the end of a Livewire test to ensure that the middleware is restored.

**Comment by donferi** [comment]
Is there a workaround for v3? Added my first livewire test and it breaks other tests with cache control headers.
